### PR TITLE
Fix reference to a released stack variable in modem_context_sprint_ip…

### LINKED ISSUE #38172: modem_context_sprint_ip_addr returns pointer to stack array
In the function modem_context_sprint_ip_addr a local array is constructed. The pointer to that structure is returned. Since this array is constructed on the stack, outside this function, this array does not exist, resulting in unexpected behavior.

File location:
drivers/modem/modem_context.c
Lines 23 - 40.

I see two options to solve this problem:
1. Either pass the array as an argument to the function, so it declared outside this function
2. Return a structure containing the array

### PR DESCRIPTION
**Pull request by ssharks**
Suggested fix for issue #38172. How can I test it?

Fixes #38172.

### CODE REVIEW

**Comment by ssharks** [comment]
Fix the build errors

**Comment by stephanosio** [comment]
Please review the [Contribution Guidelines](https://docs.zephyrproject.org/latest/contribute/index.html#commit-guidelines) and re-format the commit messages.

**Comment by ssharks** [comment]
I hope this a correct comment message

**Review by rerickson1 — CHANGES_REQUESTED** [review]
1. `drivers: modem:` repeated twice in you commit message.
2. Need a description body in your commit message
3. Need Signed-off-by in your commit body.

**Review by rerickson1 — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

**Comment by cfriedt** [comment]
@ssharks - just some whitespace changes in your commit message for CI to pass. Please modify with `commit --amend` and do a force push

File: drivers/modem/modem_context.c [hunk 1]
```diff
@@ -20,23 +20,37 @@ LOG_MODULE_REGISTER(modem_context, CONFIG_MODEM_LOG_LEVEL);
...
+		net_addr_ntop(AF_INET, &net_sin(addr)->sin_addr,
+				     buf, buf_size);
+		buf[buf_size-1] = '\0';
+		return 0;
 	}
 
 	LOG_ERR("Unknown IP address family:%d", addr->sa_family);
-	strcpy(buf, "unk");
-	return buf;
+	char unknown_str[] = "unk";
```
Comment (stephanosio) [inline drivers/modem/modem_context.c:47]: Consider moving variable declarations to the top of the scope (function).

File: drivers/modem/quectel-bg9x.c [hunk 2]
```diff
@@ -735,9 +735,20 @@ static int offload_connect(void *obj, const struct sockaddr *addr,
 
 	k_sem_reset(&mdata.sem_sock_conn);
 
+	char ip_str[NET_IPV6_ADDR_LEN];
```
Comment (stephanosio) [inline drivers/modem/quectel-bg9x.c:738]: Consider moving variable declarations to the top of the scope (function).

File: drivers/modem/ublox-sara-r4.c [hunk 3]
```diff
@@ -1574,8 +1582,17 @@ static int offload_connect(void *obj, const struct sockaddr *addr,
 		return 0;
 	}
 
+	char ip_str[NET_IPV6_ADDR_LEN];
```
Comment (stephanosio) [inline drivers/modem/ublox-sara-r4.c:1585]: Consider moving variable declarations to the top of the scope (function).

**Review by stephanosio — CHANGES_REQUESTED** [review]

File: drivers/modem/modem_context.c [hunk 4]
```diff
@@ -20,23 +20,38 @@ LOG_MODULE_REGISTER(modem_context, CONFIG_MODEM_LOG_LEVEL);
...
 	if (addr->sa_family == AF_INET6) {
-		return net_addr_ntop(AF_INET6, &net_sin6(addr)->sin6_addr,
-				     buf, sizeof(buf));
+		if (buf_size < NET_IPV6_ADDR_LEN) {
+			return -ENOMEM;
+		}
+
+		net_addr_ntop(AF_INET6, &net_sin6(addr)->sin6_addr,
+				     buf, buf_size);
+		buf[buf_size - 1] = '\0';
```
Comment (stephanosio) [inline drivers/modem/modem_context.c:34]: This should not be necessary; instead, it should check the return value of `net_addr_ntop` and return an error code upon failure in which case the content of `buf` should not be used.

File: drivers/modem/modem_context.c [hunk 5]
```diff
@@ -20,23 +20,38 @@ LOG_MODULE_REGISTER(modem_context, CONFIG_MODEM_LOG_LEVEL);
 
 static struct modem_context *contexts[CONFIG_MODEM_CONTEXT_MAX_NUM];
 
-char *modem_context_sprint_ip_addr(const struct sockaddr *addr)
+int modem_context_sprint_ip_addr(const struct sockaddr *addr, char *buf, int buf_size)
 {
-	static char buf[NET_IPV6_ADDR_LEN];
+	char unknown_str[] = "unk";
```
Comment (stephanosio) [inline drivers/modem/modem_context.c:25]:
```suggestion
	const char unknown_str[] = "unk";
```

**Comment by stephanosio** [comment]
> Does anyone has an idea why the maintainers check failed this time?

@ssharks Try rebasing onto the latest master. A few entries in the `MAINTAINERS.yml` were fixed recently.

File: drivers/modem/modem_context.h [hunk 6]
```diff
@@ -91,10 +91,12 @@ struct modem_context {
...
  *
  * @param  addr: sockaddr to be converted
+ * @param  buf:  Buffer to store IP in string form
+ * @param  buf_size:  buffer size
  *
- * @retval Buffer with IP in string form
+ * @retval 0 if ok, < 0 if error.
  */
-char *modem_context_sprint_ip_addr(const struct sockaddr *addr);
+int modem_context_sprint_ip_addr(const struct sockaddr *addr, char *buf, int buf_size);
```
Comment (stephanosio) [inline drivers/modem/modem_context.h:99]:
```suggestion
int modem_context_sprint_ip_addr(const struct sockaddr *addr, char *buf, size_t buf_size);
```

**Review by stephanosio — DISMISSED** [review]

File: drivers/modem/modem_context.h (hunk 6 above)
Comment (ssharks) [reply drivers/modem/modem_context.h]: Good suggestions, I implemented them

File: drivers/modem/modem_context.c [hunk 7]
```diff
@@ -20,23 +20,38 @@ LOG_MODULE_REGISTER(modem_context, CONFIG_MODEM_LOG_LEVEL);
...
 
 	if (addr->sa_family == AF_INET) {
-		return net_addr_ntop(AF_INET, &net_sin(addr)->sin_addr,
-				     buf, sizeof(buf));
+		if (buf_size < NET_IPV4_ADDR_LEN) {
+			return -ENOMEM;
+		}
+		net_addr_ntop(AF_INET, &net_sin(addr)->sin_addr,
+				     buf, buf_size);
+		buf[buf_size - 1] = '\0';
```
[the remark this answers is not in the document]
Comment (ssharks) [reply drivers/modem/modem_context.c]: Same

File: drivers/modem/modem_context.c (hunk 5 above)
Comment (ssharks) [reply drivers/modem/modem_context.c]: Done

File: drivers/modem/modem_context.c (hunk 4 above)
Comment (ssharks) [reply drivers/modem/modem_context.c]: Seems reasonable I just overlooked this review comment, processed it

**Comment by ssharks** [comment]
@
 
Collaborator
cfriedt ### Thanks for reminding, just overlooked it

File: drivers/modem/quectel-bg9x.c [hunk 8]
```diff
@@ -735,9 +736,18 @@ static int offload_connect(void *obj, const struct sockaddr *addr,
...
+		LOG_ERR("Closing the socket!!!");
+		socket_close(sock);
+		errno = -ret;
+		return -1;
+	}
+
 	/* Formulate the complete string. */
 	snprintk(buf, sizeof(buf), "AT+QIOPEN=%d,%d,\"%s\",\"%s\",%d,0,0", 1, sock->sock_fd, protocol,
-		 modem_context_sprint_ip_addr(addr), dst_port);
+			ip_str, dst_port);
```
Comment (stephanosio) [inline drivers/modem/quectel-bg9x.c:750]: Nit (styling issue): the beginning of this line is no longer aligned with the `(` from the previous line.
```suggestion
		 ip_str, dst_port);
```

File: drivers/modem/ublox-sara-r4.c [hunk 9]
```diff
@@ -1574,8 +1583,15 @@ static int offload_connect(void *obj, const struct sockaddr *addr,
...
+	ret = modem_context_sprint_ip_addr(addr, ip_str, sizeof(ip_str));
+	if (ret != 0) {
+		errno = -ret;
+		LOG_ERR("Error formatting IP string %d", ret);
+		return -1;
+	}
+
 	snprintk(buf, sizeof(buf), "AT+USOCO=%d,\"%s\",%d", sock->id,
-		 modem_context_sprint_ip_addr(addr), dst_port);
+			ip_str, dst_port);
```
Comment (stephanosio) [inline drivers/modem/ublox-sara-r4.c:1594]: same here

**Review by stephanosio — COMMENTED** [review]
Looks ok overall.

File: drivers/modem/quectel-bg9x.c (hunk 8 above)
Comment (ssharks) [reply drivers/modem/quectel-bg9x.c]: Fixed

File: drivers/modem/ublox-sara-r4.c (hunk 9 above)
Comment (ssharks) [reply drivers/modem/ublox-sara-r4.c]: Fixed, I also remove the spurious space

Comment (stephanosio) [reply drivers/modem/ublox-sara-r4.c]: Thanks. The space is not spurious though -- without the space, it is no longer properly aligned:
```
	snprintk(buf, sizeof(buf), "AT+USOCO=%d,\"%s\",%d", sock->id,
		 modem_context_sprint_ip_addr(addr), dst_port);
		ip_str, dst_port);
```

Comment (cfriedt) [reply drivers/modem/ublox-sara-r4.c]: Looks to be resolved, based on the Coding Guidelines checks.

**Pull request merged**